File: Tree.java
import java.util.Set;
import java.util.HashSet;
import java.util.Stack;

public class Tree{
  private Tree parent;
  private Set<Tree> children = new HashSet<Tree>();
  private String val;
  
  public Tree(String val){
    this.val = val;
  }

  public Tree(String val, Tree parent){
    this.val = val;
    this.parent = parent;
    parent.addChild(this);
  }

  public void hello(){
    this.hello("");
  }

  public void addChild(Tree child){
    this.children.add(child);
  }

  public void hello(String indent){
    System.out.println(indent + this.val);
    for(Tree child: this.children){
      child.hello(indent + "  ");
    }
  }

  public int depth(){ 
    int depth = 0;
    Stack<Tree> wq = new Stack<Tree>();
    Stack<Tree> path = new Stack<Tree>();

    wq.push(this);
    int count = 0;
    Tree r;
    while(!wq.empty()){
      count = count + 1;
      r = wq.peek();
      if(!path.empty() && r == path.peek()){
        if(path.size()>depth){
          depth = path.size();
        }
        path.pop();
        wq.pop();
      } else {
        path.push(r);
        for(Tree child: r.children){
          wq.push(child);
        }
      }
    }
    
    return depth - 1;
  }

  public static void main(String[] args){
    Tree t = new Tree("Groot");
    Tree t1 = new Tree("c1", t);
    Tree t2 = new Tree("c2", t);
    Tree t11 = new Tree("c11", t1);
    t.hello();
    System.out.println("Depth: " + t.depth());
    
  }
}

File: ChallengeHelper/Tree.py
class Tree:
    def __init__(self, value, parent=None):
        self.value = value
        self.parent = parent
        if(parent):
            parent.add_child(self)
        self.children = set()

    def set_parent(self, parent):
        self.parent = parent;

    def val(self):
        return self.value

    def add_child(self, child):
        self.children.add(child)

    def hello(self, indent=""):
        print(indent + self.val())
        for child in self.children:
            child.hello(indent + "  ")

    def depth_rec(self):
        if len(self.children) == 0:
            return 0
        else:
            return 1 + max([child.depth() for child in self.children])

    def depth(self):
        depth = 0
        wq = []
        path = []

        wq.append(self)
        count = 0
        while wq:
            count = count + 1
            #print([n.val() for n in wq], [n.val() for n in path])
            r = wq[-1]
            if path and r == path[-1]:
                if len(path) > depth:
                    depth = len(path)
                path.pop()
                wq.pop()
            else:
                path.append(r)
                for child in r.children:
                    wq.append(child)

        return depth - 1
        

